### PR TITLE
Correctly pruning unused definitions in metaschema composition

### LINKED ISSUE #180: Compiling stylesheet fails: unknown accumulator Q{}definition-count
While compiling a stylesheet, used in the schema and converter generation, to a .SEF file, received the following error message: 

Failed to compile stylesheet: Static error in XPath on line 38 in compose/metaschema-prune-unused-definitions.xsl {accumulator-after('definition-count')}: Unknown accumulator Q{}definition-count
Error Q{http://www.w3.org/2005/xqt-errors}XTDE3340 at xpath.xsl#963
    Failed to compile stylesheet:
Error Q{http://www.w3.org/2005/xqt-errors}XTDE3340 at xpath.xsl#963
    Static error in XPath on line 38 in compose/metaschema-prune-unused-definitions.xsl {accumulator-after('definition-count')}: Unknown accumulator Q{}definition-count

This could be a Saxon JS bug as well. The traceback reports an error in an accumulator. 

This could possibly affect the transformation process to schemas and converters.

This occurs when the following command is executed:  "xslt3 -t -xsl:metaschema-prune-unused-definitions.xsl -export:metaschema-prune-unused-definitions.sef.json -nogo"

The stylesheet being compiled is located here: https://github.com/usnistgov/metaschema/blob/54d1d8d24edb0f07fe0e8effd5ea476e890ecdb7/toolchains/xslt-M4/compose/metaschema-prune-unused-definitions.xsl

### PR DESCRIPTION
**Pull request by wendellpiez**
By replacing the pruning step in metaschema composition, this closes #180 and #198.

Also we no longer need #196 with this PR (right @aj-stein-nist ?)

Also includes plenty of good work towards #81 #186, #201.

### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/metaschema/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/metaschema/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Do all automated CI/CD checks pass?

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you included examples of how to use your new feature(s)?
- [x] Have you updated all website](https://pages.nist.gov/metaschema) and readme documentation affected by the changes you made? Changes to the website can be made in the website/content directory of your branch.

### CODE REVIEW

**Comment by wendellpiez** [comment]
This PR now contains a solution, with running unit tests, to expose and remedy #198.

I also suggest we try results from this pipeline applied to OSCAL metaschemas, to head off potential future unpleasantness.

**Comment by aj-stein-nist** [comment]
Hey @wendellpiez, Dave was reviewing PRs with me and asked for a brief on #196 and reminded him the work is going to continue here. Is this issue ready for review and merge or are you still working on it?

**Comment by wendellpiez** [comment]
@david-waltermire-nist and @aj-stein-nist - this issue as scoped above is done: this PR contains a new XSLT step for pruning that repairs the functional bug, plus a unit test that demonstrates the solution. This addresses both #180 and #198.

It also includes lots of other efforts (Metaschema unit tests) described in other Issues, so it becomes a dependency for them. Please pull whenever convenient!

**Comment by aj-stein-nist** [comment]
> @david-waltermire-nist and @aj-stein-nist - this issue as scoped above is done: this PR contains a new XSLT step for pruning that repairs the functional bug, plus a unit test that demonstrates the solution. This addresses both #180 and #198.

Sounds good. Shall we meet and get it rebased and good to go for final review by Dave?

**Comment by aj-stein-nist** [comment]
I am going to review it as-is, @david-waltermire-nist, but @wendellpiez and I discussed this yesterday afternoon and we wanted confirmation: current sprint work like this is target `master` and not the `develop` branch, correct? I compared to the two branches and more urgent pressing work is being merged into `master`, so I presume this will need to get updated, but want to make sure.

File: test-suite/metaschema-xspec/metaschema-test-helper.xsl [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        catalog | metadata | param | control | group | back-matter | prop | link | part | usage | constraint | guideline | select | remarks | description | test | revisions | revision | role | location | party | responsible-party | address | external-id | resource | citation | rlink | base64
+        
+        
+    -->
+    <!--<xsl:template match="/">
+        <xsl:param name="source" select="." as="document-node()"/>
+        <xsl:document>
+          <xsl:apply-templates select="$source" mode="scrubbing"/>
+        </xsl:document>
+    </xsl:template>-->
```
Comment (aj-stein-nist) [inline test-suite/metaschema-xspec/metaschema-test-helper.xsl:42]: Minor: should we not just delete this because it is dead code?

```suggestion
```

File: test-suite/metaschema-xspec/readme.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+### NOT CURRENTLY FUNCTIONAL
+
+* XSpec `metaschema-basic-schematron.xspec` testing Schematron:  `../../toolchains/xslt-M4/validate/metaschema-simple-check.sch`
+
+*Needs updating and pointing to Schematron in its new location.*
+
+* XSpec `metaschema-composition-schematron.xspec` testing Schematron: `../../toolchains/xslt-M4/validate/metaschema-composition-check.sch`
+
+*Needs updating and pointing to Schematron in its new location*
+
```
Comment (aj-stein-nist) [inline test-suite/metaschema-xspec/readme.md:37]: Do we want to add this to documentation, and not just add it instead to the epic where we will fix it?

File: test-suite/metaschema-xspec/schematron/metaschema-composition-schematron.xspec [hunk 3]
```diff
@@ -1,14 +1,15 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <x:description xmlns:x="http://www.jenitennison.com/xslt/xspec"
     xmlns:m="http://csrc.nist.gov/ns/oscal/metaschema/1.0"
-    schematron="../../toolchains/xslt-M4/validate/metaschema-composition-check.sch">
-    <x:
-    <x:param name="global-context" select="document('scope-local-importing_metaschema.xml')"/>
+    schematron="../../../toolchains/xslt-M4/validate/metaschema-composition-check.sch"
+    run-as="external">
+    
+    <!--<x:param name="global-context" select="document('scope-local-importing_metaschema.xml')"/>-->
```
Comment (aj-stein-nist) [inline test-suite/metaschema-xspec/schematron/metaschema-composition-schematron.xspec:7]: Minor: consider deleting if unused?

```suggestion
```

File: toolchains/xslt-M4/compose/metaschema-collect.xsl [hunk 4]
```diff
@@ -66,7 +66,7 @@
         <xsl:copy>
             <xsl:attribute name="abstract" select="$metaschema-abstract-default"/>
             <xsl:copy-of select="@* except @xsi:*"/>
-            <xsl:attribute name="module" select="short-name"/>
+            <xsl:attribute name="module" select="(@module[matches(.,'\S')],short-name)[1]"/>
```
Comment (aj-stein-nist) [inline toolchains/xslt-M4/compose/metaschema-collect.xsl:69]: I know we have tested the pruning code, but does this change to collect have significant implications? I will test in the OSCAL repo, but the OSCAL metaschemas make no use of `module`, but this concatenates to it will still work in the absence of a `@module` that matches the name of the `.` context, correct?

**Review by aj-stein-nist — DISMISSED** [review]
Looks good, I have mostly minor feedback and one question about a modification to the Metaschema collection part of the composition pipeline. Other than that, nice work and thanks for cleaning up the repo!

File: test-suite/metaschema-xspec/metaschema-test-helper.xsl (hunk 1 above)
Comment (aj-stein-nist) [reply test-suite/metaschema-xspec/metaschema-test-helper.xsl:42]: Weird I am not sure why it gets highlighted to be like it is not commented code in the suggestion diff, but it is! haha

File: toolchains/xslt-M4/compose/metaschema-collect.xsl (hunk 4 above)
Comment (aj-stein-nist) [reply toolchains/xslt-M4/compose/metaschema-collect.xsl:69]: Tested current OSCAL and it is good, so I surpress this comment in my feedback. Feel free to ignore it!

**Comment by aj-stein-nist** [comment]
Per our conversation earlier this afternoon, I was able [to make a branch on my fork of OSCAL and test the output line by line](https://github.com/aj-stein-nist/OSCAL/commit/029363db60ab77400c5ff35d71ef8b3cc3fcc3a4) just to make double sure. It looks good, line-for-line parity with all models, couldn't get anything to commit on my end.

File: test-suite/metaschema-xspec/readme.md (hunk 2 above)
Comment (wendellpiez) [reply test-suite/metaschema-xspec/readme.md:37]: I think both. (Making slight adjustment accordingly.)

File: test-suite/metaschema-xspec/schematron/metaschema-composition-schematron.xspec [hunk 5]
```diff
@@ -1,14 +1,15 @@
...
     <x:param name="allow-foreign">true</x:param>
 
     <!-- Basic starter XSpec shows both callout and inline scenarios. -->
     <x:scenario label="poc-on-remote-file">
-        <x:context href="scope-local-importing_metaschema.xml"/><!--
+        <x:context  href="scope-local-importing_metaschema.xml"/><!--
             <METASCHEMA xmlns="http://csrc.nist.gov/ns/oscal/metaschema/1.0">
                 <short-name>boohoo</short-name>
             </METASCHEMA>
```
[the remark this answers is not in the document]
Comment (wendellpiez) [reply test-suite/metaschema-xspec/schematron/metaschema-composition-schematron.xspec:15]: Actually this stubs out syntax we will be wanting so I'm leaving it for now (good idea in general though).

File: toolchains/xslt-M4/compose/readme.md [hunk 6]
```diff
@@ -2,49 +2,57 @@ Transformations in this subdirectory support Metaschema composition, that is the
 
 Generally, metaschema composition will be a necessary preliminary to producing a schema, documentation or analysis of a single metaschema.
```
[the remark this answers is not in the document]
Comment (wendellpiez) [reply toolchains/xslt-M4/compose/readme.md:3]: Actually, "preliminary" can be a noun ... https://www.merriam-webster.com/dictionary/preliminary

File: toolchains/xslt-M4/compose/testing/readme.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+With these in place, each step can have at least one test that works on inputs in the `expected` folder of the preceding branch, to help maintain alignment among the tests.
+
+TODO:
+- update present tests
+- build out more tests
+  - work up to three variants? skeletal, basic and pathological (showing errors/exceptions)
+  - also unit test any special handling or reasonable edge conditions 
+- pull together 'predecessor' tests that formalize the chain
+- extend XSD / Schematron to support final results?
```
[the remark this answers is not in the document]
Comment (wendellpiez) [reply toolchains/xslt-M4/compose/testing/readme.md:25]: I am hoping that we can work over all the unit testing code enough such that old TODO notices are not a *great* source of embarrassment. I am not opposed to moving this to an Issue - but here it is.

**Comment by wendellpiez** [comment]
Next committing an update addressing points of discussion above.

File: test-suite/metaschema-xspec/schematron/metaschema-composition-schematron.xspec (hunk 5 above)
[the remark this answers is not in the document]
Comment (aj-stein-nist) [reply test-suite/metaschema-xspec/schematron/metaschema-composition-schematron.xspec:15]: Fine by me, I will close this comment out then.

File: toolchains/xslt-M4/compose/readme.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (aj-stein-nist) [reply toolchains/xslt-M4/compose/readme.md:3]: Fine by me to leave it as-is, will close the comment.

**Comment by aj-stein-nist** [comment]
> I am going to review it as-is, @david-waltermire-nist, but @wendellpiez and I discussed this yesterday afternoon and we wanted confirmation: current sprint work like this is target `master` and not the `develop` branch, correct? I compared to the two branches and more urgent pressing work is being merged into `master`, so I presume this will need to get updated, but want to make sure.

@wendellpiez sorry for the delay on this but I talked to Dave and he said he wanted this one pointing to the `develop` branch.

File: toolchains/xslt-M4/compose/testing/readme.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (aj-stein-nist) [reply toolchains/xslt-M4/compose/testing/readme.md:25]: While doing triage by myself review usnistgov/metaschema#201 and updated the AC to reflect this list.

File: test-suite/metaschema-xspec/readme.md (hunk 2 above)
Comment (aj-stein-nist) [reply test-suite/metaschema-xspec/readme.md]: WIll resolve this part as is covered by other updates to usnistgov/metaschema#201.

**Review by david-waltermire-nist — APPROVED** [review]
Looks good to me.
